File: 541-Reverse String II-反转字符串 II/541-Reverse String II-反转字符串 II.java
class Solution {
    public String reverseStr(String s, int k) {
        char[] sArr = s.toCharArray();
    	int start = 0;
    	int end = 0;
    	for(int i = 0; i < sArr.length; i += 2 * k) {
    		start = i;
    		/*关键：确定尾部指针的位置
    		sArr.length - 1：剩余的数量
    		start + k + 1：k+1个数
    		实际上最后剩余数num->[k, 2k]的操作方式普通的操作方式相同，所以实际上只有两种情况
    		*/
    		end = Math.min(sArr.length - 1, start + k - 1);
    		while(start < end) {
    			//学到了
    			sArr[start] ^= sArr[end];
    			sArr[end] ^= sArr[start];
    			sArr[start] ^= sArr[end];
                start++;
                end--;
    		}
    	}
    	return new String(sArr);
    }
}

File: 541-Reverse String II-反转字符串 II/541-Reverse String II-反转字符串 II.js
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
 var reverseStr = function(s, k) {
    let sArr = s.split('')
    let len = s.length
    let i
    //每计数至 2k 个字符，就反转这 2k 字符中的前 k 个字符。
    for(i = 0; i < s.length; i = i + 2 * k){
        revertPart(sArr, i, i + k)
    }
    
    if(len - i < k){                    //如果剩余字符少于 k 个，则将剩余字符全部反转。
        revertPart(sArr, i, len)
    }else{                              //如果剩余字符小于 2k 但大于或等于 k 个，则反转前 k 个字符，其余字符保持原样。
        revertPart(sArr, i, i + k)
    }
    return sArr.join('')
};

var revertPart = function(sArr, start, end){
    let temp
    let count = end + start - 1
    for(let j = start; j < count / 2; j++){
        temp = sArr[j]
        sArr[j] = sArr[count - j]
        sArr[count - j] = temp
    }
}